pkg/compactor/shard_aware_deduplicate_filter.go

```diff
@@ -276,12 +285,19 @@ func (b *blockWithSuccessors) isFullyIncludedInSuccessors() bool {
 	return uint64(len(shards)) == shardCount
 }
 
-func (b *blockWithSuccessors) getDuplicateBlocks(output map[ulid.ULID]struct{}) {
-	if b.meta != nil && b.isFullyIncludedInSuccessors() {
-		output[b.meta.ULID] = struct{}{}
-	}
+func (b *blockWithSuccessors) getDuplicateBlocks() map[ulid.ULID]struct{} {
+	var getDuplicates func(b *blockWithSuccessors, output map[ulid.ULID]struct{})
```

Comment:
Why is this declaring a closure? Seems like a plain old for-loop would work.

Reply:
@56quarters getDuplicateBlocks() is a function that returns a set of ULIDs of the blocks that are duplicates. In order to determine all those block, we need the closure, getDuplicates, that takes as input a block, and the current set of duplicates, and that finds all the duplicates by recursively calling successors of the input block.

At the bottom of the getDuplicateBlocks, an empty map is initialized, passed as parameter to the closure, and then it is returned as a result. 

`func (b *blockWithSuccessors) getDuplicateBlocks() map[ulid.ULID]struct{} {
	var getDuplicates func(b *blockWithSuccessors, output map[ulid.ULID]struct{})
	getDuplicates = func(b *blockWithSuccessors, output map[ulid.ULID]struct{}) {
		if b.meta != nil && b.isFullyIncludedInSuccessors() {
			output[b.meta.ULID] = struct{}{}
		}

		for _, s := range b.successors {
			getDuplicates(s, output)
		}
	}

	duplicateULIDs := make(map[ulid.ULID]struct{})
	getDuplicates(b, duplicateULIDs)
	return duplicateULIDs
}`